### ISSUE TITLE
[Subshell] Support for ash + bugfixes for bash, fish

### ISSUE THREAD

**Issue by mc-butler** [MidnightCommander/mc#2742]
> [!IMPORTANT]
> This issue was migrated from Trac:
>
> | **Origin** | https://midnight-commander.org/ticket/2742 |
> | :------------- | :------------------------------------------------------------------------------------------------ |
> | **Reporter** | `kriegaex` (Alexander@….name) |
> | **Mentions** | `pahan@….info` (@Hubbitus) |

<p>
This patch is related to a <a class="ext-link" href="http://mail.gnome.org/archives/mc-devel/2012-March/msg00000.html">thread on mc-devel</a>. I am copying the main part of my <a class="ext-link" href="http://mail.gnome.org/archives/mc-devel/2012-March/msg00008.html">previous post</a> to this ticket because I was asked to create one:<br />
</p>
<blockquote class="citation">
<p>
I have resolved the remaining issues with differentiating between <em>BusyBox ash</em> and <em>Debian ash (dash)</em>. I also refactored the subshell code and documentation a bit for all subshell types. I installed and tested <strong>all</strong> of them on my Ubuntu 11.10 64-bit desktop system plus <em>Bash</em> and <em>BusyBox ash</em> on my embedded mipsel platform.<br />
</p>
<p>
I think now I have reached a mature enough code state for you to commit upstream because everything which worked before still does and several things which were buggy or unavailable before have been fixed/added.<br />
</p>
<p>
The patches currently contain the following changes:<br />
</p>
<ul><li><em>030-bash_inputrc.patch</em>
<ul><li>Fix non-functional <tt>INPUTRC</tt> for <em>bash</em> (variable was unset and never used)
</li></ul></li></ul><ul><li><em>040-ash_as_subshell.patch</em>
<ul><li>Fix chdir for <em>fish</em>: communication with subshell was not working and showing an error message because <tt>pwd</tt> returns abbreviated home directory "~" instead of full path name. So now we use <tt>$PWD</tt> and it works. This must have been broken for a while. This is unrelated to <em>ash</em>, but I included it in the refactoring patch for <em>src/subshell.c</em> anyway because it is small and the same code section was also changed for other reasons (set <em>fish</em> prompt).
</li><li>All subshells now have a dynamic, meaningful prompt <tt>user@host:cwd</tt>. This was tricky for <em>ash/dash</em> and a little easier for the other shells.
</li><li>Shell type recognition for automatic subshell selection has been improved: Now not only the <tt>SHELL</tt> names are checked, but also their link targets in order to determine the correct type (e.g. if a <em>ksh</em> is really a <em>zsh</em> or if an <em>ash</em> is <em>BusyBox</em> or <em>dash</em>, if <em>sh</em> is whichever type or if <em>bash</em> is symlinked to <em>BusyBox</em> (but <em>BB</em> still is just an <em>ash</em>, never a <em>bash</em>). Even if the type could not be determined or a compatible subshell not found via inspection of <tt>SHELL</tt> and <em>/etc/passwd</em>, function <tt>OS_Setup</tt> in <em>src/main.c</em> now searches for known shells by their usual paths. For instance, if you have an exotic login shell called <em>foosh</em>, you can now have subshell support anyway, if MC finds an <em>ash</em>, <em>bash</em> etc. on the system.
</li><li>I have added many more comments to the code sections I changed. Maybe some of you think there are too many now, but I think this tricky part of the code base should be documented rather too much than not enough, so as to make it easier for maintainers to upgrade subshell support or add new subshell types in the future. Sometimes it is not enough to have working code, but helpful to understand <strong>why</strong> the code was written in a certain way.
</li></ul></li></ul><ul><li><em>040-ash_as_subshell_additional.patch</em>
<ul><li><em>ashrc</em> init file added to config file list
</li><li>Subshell documentation page updated with a few corrections, better formatting and some new information (<em>doc/man/mc.1.in</em>), e.g. how to use init files for <em>bash</em> and <em>ash</em>, how to manually call <em>mc</em> with <tt>SHELL</tt> variable set to another shell type if you want to have a specific subshell type.
</li></ul></li></ul><p>
BTW, the patches have been created for mc-4.8.1, <strong>not</strong> for the trunk/master. But they should be easy to merge into the master.<br />
</p>
<p>
All in all I believe that adding these patches to MC will significantly improve subshell support, fixing two old issues and adding new subshell types often used on servers or embedded systems. That could make MC even more popular. E.g. I have heard from several people that they would like to get rid of feature-rich, but big shells like <em>bash</em>, <em>zsh</em> or <em>tcsh</em> on such systems if they had <em>BusyBox</em> or <em>dash</em> subshell support. I think <em>MC</em> is far from dead and did what I could with my humble means to improve a small part of it.<br />
</p>
<p>
Open issues (legacy, not created by my patches):<br />
</p>
<ul><li>Still open is the old issue of <em>fish</em> not reliably showing its prompt on the bottom line of <em>mc</em>. With my limited knowledge I could not find out why this happens, but I noticed an interesting fact: If I start <em>mc</em> via <em>strace</em> (I did this because I was debugging/logging something during development), all of a sudden the <em>fish</em> prompt is shown reliably on my test machine. Maybe <em>strace</em> was slowing down something, so this might be a timing issue, possibly a race condition, but this is just a guess. It might be a starting point for further investigation by one of you guys.
</li><li>Another (old) open issue which still exists is that if you cd to another directory in full-screen mode (hide panels with <em>Ctrl-O</em>), then show the panels again with <em>Ctrl-O</em>, the new path is not immediately shown on the bottom line, but only after some user interaction such as pressing Tab twice, switching back and forth between left/right panel.
</li></ul></blockquote>

> [!NOTE]
> Original attachments:
>
> * [030-bash_inputrc.patch](https://github.com/MidnightCommander/trac-archive/blob/master/attachments/ticket/2742/030-bash_inputrc.patch) ([raw](https://github.com/MidnightCommander/trac-archive/raw/master/attachments/ticket/2742/030-bash_inputrc.patch)) by `kriegaex` (Alexander@….name) on <span title="2012-03-08T11:30:56+00:00">`Mar 8, 2012 at 11:30 UTC`</span>
> * [040-ash_as_subshell.patch](https://github.com/MidnightCommander/trac-archive/blob/master/attachments/ticket/2742/040-ash_as_subshell.patch) ([raw](https://github.com/MidnightCommander/trac-archive/raw/master/attachments/ticket/2742/040-ash_as_subshell.patch)) by `kriegaex` (Alexander@….name) on <span title="2012-03-08T11:32:02+00:00">`Mar 8, 2012 at 11:32 UTC`</span>
> * [040-ash_as_subshell_additional.patch](https://github.com/MidnightCommander/trac-archive/blob/master/attachments/ticket/2742/040-ash_as_subshell_additional.patch) ([raw](https://github.com/MidnightCommander/trac-archive/raw/master/attachments/ticket/2742/040-ash_as_subshell_additional.patch)) by `kriegaex` (Alexander@….name) on <span title="2012-03-08T11:33:03+00:00">`Mar 8, 2012 at 11:33 UTC`</span>
> * [030-bash_inputrc___mc-4.8.8.patch](https://github.com/MidnightCommander/trac-archive/blob/master/attachments/ticket/2742/030-bash_inputrc___mc-4.8.8.patch) ([raw](https://github.com/MidnightCommander/trac-archive/raw/master/attachments/ticket/2742/030-bash_inputrc___mc-4.8.8.patch)) by `er13` (ernews@….de) on <span title="2013-04-06T13:57:26+00:00">`Apr 6, 2013 at 13:57 UTC`</span>
> * [040-ash_as_subshell___mc-4.8.8.patch](https://github.com/MidnightCommander/trac-archive/blob/master/attachments/ticket/2742/040-ash_as_subshell___mc-4.8.8.patch) ([raw](https://github.com/MidnightCommander/trac-archive/raw/master/attachments/ticket/2742/040-ash_as_subshell___mc-4.8.8.patch)) by `er13` (ernews@….de) on <span title="2013-04-06T13:57:36+00:00">`Apr 6, 2013 at 13:57 UTC`</span>
> * [050-ash_as_subshell_additional___mc-4.8.8.patch](https://github.com/MidnightCommander/trac-archive/blob/master/attachments/ticket/2742/050-ash_as_subshell_additional___mc-4.8.8.patch) ([raw](https://github.com/MidnightCommander/trac-archive/raw/master/attachments/ticket/2742/050-ash_as_subshell_additional___mc-4.8.8.patch)) by `er13` (ernews@….de) on <span title="2013-04-06T13:57:45+00:00">`Apr 6, 2013 at 13:57 UTC`</span>
> * [0001-Ticket-2742-detect-csh-as-tcsh-by-name.patch](https://github.com/MidnightCommander/trac-archive/blob/master/attachments/ticket/2742/0001-Ticket-2742-detect-csh-as-tcsh-by-name.patch) ([raw](https://github.com/MidnightCommander/trac-archive/raw/master/attachments/ticket/2742/0001-Ticket-2742-detect-csh-as-tcsh-by-name.patch)) by `zaytsev-work` (@zyv) on <span title="2016-03-16T14:00:13+00:00">`Mar 16, 2016 at 14:00 UTC`</span>
